hirefire/procs/celery.py

```diff
@@ -64,4 +64,21 @@ def quantity(self):
         """
         Returns the aggregated number of tasks of the proc queues.
         """
-        return sum([self.channel._size(queue) for queue in self.queues])
+        if hasattr(self.channel, '_size'):
+            # Redis
+            return sum(self.channel._size(queue) for queue in self.queues)
+        # AMQP
+        try:
+            from librabbitmq import ChannelError
+        except ImportError:
+            from amqp.exceptions import ChannelError
+        count = 0
+        for queue in self.queues:
+            try:
+                queue = self.channel.queue_declare(queue, passive=True)
```

Comment:
> On 21 Apr 2015, at 10:02, Marc Tamlyn <notifications@github.com> wrote:
> 
> In hirefire/procs/celery.py:
> 
> > @@ -64,4 +64,21 @@ def quantity(self):
> >          """
> >          Returns the aggregated number of tasks of the proc queues.
> >          """
> > -        return sum([self.channel._size(queue) for queue in self.queues])
> > +        if hasattr(self.channel, '_size'):
> > +            # Redis
> > +            return sum(self.channel._size(queue) for queue in self.queues)
> > +        # AMQP
> > +        try:
> > +            from librabbitmq import ChannelError
> > +        except ImportError:
> > +            from amqp.exceptions import ChannelError
> > +        count = 0
> > +        for queue in self.queues:
> > +            try:
> > +                queue = self.channel.queue_declare(queue, passive=True)
> 
> Looks like it does actually. On redis it will return a queue_declare_ok_t named tuple from https://github.com/celery/py-amqp/blob/master/amqp/protocol.py which will have a message_count attribute.
> 
> The behaviour on non existent queues with passive=True does seem to vary, but we could likely write a clean enough implementation here with just except Exception which seems like it should work in all cases. Most of the non-amqp transport backends use a subclass of the virtual channel to mimic the behaviour of the amqp transport backends.

Hm, blank exceptions aren’t awesome and would make things hard to debug. So if you can find an alternative, that’d be preferred. Have you gotten feedback from Ask?

> One drawback of this approach, it does more work than strictly necessary for all of the virtual backends. I believe all of those support the _size method, so that can be used directly. This reduces the amount of communication we need to do - instead of "get this queue if exists and find out how bit it as an how many active workers there are", we just say "get the size of this queue". For AMQP I believe the latter independently is not a possible operation, and you get the size of the queue for free when testing for its existence. For other backends this could require multiple calls across the connection. I don't think performance is too major an issue here, but it's worth considering.

Well given the amount of requests hirefire sends we can predict how much extra connections and traffic this would create. At the very least we should document that prominently so users aren’t surprised about what happens.

What I’m said about is that this isn’t well defined in kombu despite the fact that it wraps to drastically different transport systems. Maybe you should ask for that feature?